e2e2/test/cases/nvidia/mpi_test.go

```diff
@@ -160,3 +191,164 @@ func mpiJobSucceeded(obj k8s.Object) bool {
+	restConfig := cfg.Client().RESTConfig()
+	// Initialize Kubernetes client from the provided REST config
+	clientset, err := kubernetes.NewForConfig(restConfig)
+	if err != nil {
+		return nil, fmt.Errorf("failed to create Kubernetes client: %v", err)
+	}
+
+	// List nodes in the cluster
+	nodes, err := clientset.CoreV1().Nodes().List(context.TODO(), metav1.ListOptions{})
+	if err != nil {
+		return nil, fmt.Errorf("failed to list nodes: %v", err)
+	}
+
+	// Ensure at least one node is available
+	if len(nodes.Items) == 0 {
+		return nil, fmt.Errorf("no nodes found in the cluster")
+	}
+
+	// Get instance type and metadata from the first node
+	node := nodes.Items[0]
+	osType := node.Status.NodeInfo.OSImage
+
+	// Get k8s cluster version
+	discoveryClient, err := discovery.NewDiscoveryClientForConfig(restConfig)
+	if err != nil {
+		return nil, fmt.Errorf(" error in discoveryClient %v", err)
+	}
+	information, err := discoveryClient.ServerVersion()
+	if err != nil {
+		return nil, fmt.Errorf("failed to get k8s version: %v", err)
+	}
+	k8sVersion := fmt.Sprintf("%s.%s", information.Major, information.Minor)
+
+	// get ami id
+	instanceId := regexp.MustCompile(`i-[a-zA-Z0-9]+`).FindString(node.Spec.ProviderID)
+	// describe the instance
+	ec2Client := ec2.NewFromConfig(awsCfg)
+	instance, err := ec2Client.DescribeInstances(ctx, &ec2.DescribeInstancesInput{
+		InstanceIds: []string{instanceId},
+	})
+	if err != nil {
+		return nil, fmt.Errorf("failed to describe instance: %v", err)
+	}
+	amiId := *instance.Reservations[0].Instances[0].ImageId
+	//get ami name from id
+	ami, err := ec2Client.DescribeImages(ctx, &ec2.DescribeImagesInput{
+		ImageIds: []string{amiId},
+	})
+	if err != nil {
+		return nil, fmt.Errorf("failed to describe ami: %v", err)
+	}
+	amiName := ami.Images[0].Name
+	// Construct the metadata labels
+	metadataLabels := map[string]string{
+		"instance_type":      *nodeType,
+		"node_count":         fmt.Sprintf("%d", nodeCount),
+		"kubernetes_version": k8sVersion,
+		"efa_enabled":        fmt.Sprintf("%t", *efaEnabled),
+		"efa_count":          fmt.Sprintf("%d", efaPerNode),
+		"ami_id":             amiId,
+		"ami_name":           *amiName,
+		"os_type":            osType,
+	}
+
+	// Create a job to fetch the logs of meta info
+	job := &batchv1.Job{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      "metadata-job",
+			Namespace: "default",
+		},
+		Spec: batchv1.JobSpec{
+			Template: v1.PodTemplateSpec{
+				Spec: v1.PodSpec{
+					RestartPolicy: v1.RestartPolicyNever,
+					Containers: []v1.Container{
+						{
+							Name:            "metadata-job",
+							Image:           *nvidiaTestImage,
+							ImagePullPolicy: v1.PullAlways,
+							Resources: v1.ResourceRequirements{
+								Limits: v1.ResourceList{
+									"nvidia.com/gpu":        node.Status.Capacity["nvidia.com/gpu"],
+									"vpc.amazonaws.com/efa": node.Status.Capacity["vpc.amazonaws.com/efa"],
+								},
+							},
+						},
+					},
+				},
+			},
+		},
+	}
```

Comment:
I think we can use a template here to reduce the function size

Reply:
Updated in new rev